pfcli/service/client/checkpoint.py

```diff
@@ -149,41 +148,39 @@ def abort_mpu(self, ckpt_form_id: UUID, path: str, upload_id: str) -> None:
         )
 
     def _multipart_upload_file(
-        self, ckpt_form_id: UUID, file_path: str, url_dict: Dict[str, Any], ctx: tqdm
+        self,
+        ckpt_form_id: UUID,
+        file_path: str,
+        url_dict: Dict[str, Any],
+        ctx: tqdm,
+        max_workers: int,
     ) -> None:
-        # TODO (ym): parallelize each part upload.
         parts = []
         upload_id = url_dict["upload_id"]
         object_path = url_dict["path"]
+        upload_urls = url_dict["upload_urls"]
+        total_num_parts = len(upload_urls)
         try:
-            with open(file_path, "rb") as f:
-                fileno = f.fileno()
-                total_file_size = os.fstat(fileno).st_size
-                for idx, url_info in enumerate(url_dict["upload_urls"]):
-                    cursor = idx * S3_MPU_PART_MAX_SIZE
-                    f.seek(cursor)
-                    # Only the last part can be smaller than ``S3_MPU_PART_MAX_SIZE``.
-                    chunk_size = min(S3_MPU_PART_MAX_SIZE, total_file_size - cursor)
-                    wrapped_object = CustomCallbackIOWrapper(
-                        ctx.update, f, "read", chunk_size
-                    )
-                    s = Session()
-                    req = Request("PUT", url_info["upload_url"], data=wrapped_object)
-                    prep = req.prepare()
-                    prep.headers["Content-Length"] = str(chunk_size)
-                    response = s.send(prep)
-                    response.raise_for_status()
-
-                    etag = response.headers["ETag"]
-                    parts.append(
-                        {
-                            "etag": etag,
-                            "part_number": url_info["part_number"],
-                        }
+            # NOTE: excessive concurrency may results in "No buffer space available" error.
+            with ThreadPoolExecutor(max_workers=max_workers) as executor:
+                futs = [
+                    executor.submit(
+                        upload_part,
+                        file_path=file_path,
+                        chunk_index=idx,
+                        part_number=url_info["part_number"],
+                        upload_url=url_info["upload_url"],
+                        ctx=ctx,
+                        is_last_part=(idx == total_num_parts - 1),
                     )
-                assert not f.read(
-                    S3_MPU_PART_MAX_SIZE
-                ), "Some parts of your data is not uploaded. Please try again."
+                    for idx, url_info in enumerate(upload_urls)
+                ]
+                wait(futs, return_when=FIRST_EXCEPTION)
+                for fut in futs:
+                    exc = fut.exception()
+                    if exc is not None:
+                        raise exc
```

Comment:
We need to summarize the exception text and gracefully exit with an error code.

Reply:
+) Do we need `try-except` to catch `FileNotFoundError`?
I think just stringifying `exc` might be enough.

Reply:
> We need to summarize the exception text and gracefully exit with an error code.

What do you mean by "summarize the exception"? I added an `except` block to catch `KeyboardInterrupt` to handle the upload abortion by users.